src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1259,37 +1236,46 @@ internal async ValueTask<NpgsqlDataReader> ExecuteReader(CommandBehavior behavio
-                        var sendTask = NonMultiplexingWriteWrapper(connector, async, finalCt);
+                        var sendTask = NonMultiplexingWriteWrapper(connector, async, writeCt);
 
-                        // The following is a hack. It raises an exception if one was thrown in the first phases
-                        // of the send (i.e. in parts of the send that executed synchronously). Exceptions may
-                        // still happen later and aren't properly handled. See #1323.
-                        if (sendTask.IsFaulted)
-                            sendTask.GetAwaiter().GetResult();
+                        try
+                        {
+                            // The following is a hack. It raises an exception if one was thrown in the first phases
+                            // of the send (i.e. in parts of the send that executed synchronously). Exceptions may
+                            // still happen later and aren't properly handled. See #1323.
+                            if (sendTask.IsFaulted)
+                                sendTask.GetAwaiter().GetResult();
+                        }
+                        catch (OperationCanceledException e)
+                        {
+                            // User requested the cancellation - breaking the connection, as WriteBuffer doesn't do it
+                            throw connector.Break(e);
+                        }
 
                         // TODO: DRY the following with multiplexing, but be careful with the cancellation registration...
                         var reader = connector.DataReader;
                         reader.Init(this, behavior, _statements, sendTask);
                         connector.CurrentReader = reader;
                         if (async)
-                            await reader.NextResultAsync(finalCt);
+                            await reader.NextResultAsync(readCt);
                         else
                             reader.NextResult();
                         return reader;
                     }
-                    catch
+                    catch (Exception e)
                     {
                         connector.CurrentReader = null;
                         conn.Connector?.EndUserAction();
+
+                        if (connector.UserCancellationRequested && e is PostgresException pg && pg.SqlState == PostgresErrorCodes.QueryCanceled)
+                            throw new OperationCanceledException("Query was cancelled", e, cancellationToken);
+
                         throw;
                     }
                     finally
                     {
-                        if (registration.HasValue)
-                        {
-                            registration.Value.Dispose();
-                            connector.CommandCts.Stop();
-                        }
+                        registration?.Dispose();
+                        readCts.Stop();
```

Comment:
The previous code had only one check on registration, so it was a bit more efficient when there's no cancellation token.